deltacat/compute/compactor/steps/repartition.py

```diff
@@ -0,0 +1,132 @@
+        partition_ranges: List = repartition_args["ranges"]
+        task_id = get_current_ray_task_id()
+        worker_id = get_current_ray_worker_id()
+        with memray.Tracker(
+            f"repartition_{worker_id}_{task_id}.bin"
+        ) if enable_profiler else nullcontext():
+            tables: List[pa.Table] = deltacat_storage.download_delta(
+                annotated_delta,
+                storage_type=StorageType.LOCAL,
+                file_reader_kwargs_provider=read_kwargs_provider,
+            )
+            partitioned_tables = {i: [] for i in range(len(partition_ranges) + 1)}
+            total_record_count = 0
+            col_name_int64 = f"{column}_int64"
+            for table in tables:
+                total_record_count += len(table)
+                table_new = table.add_column(
+                    0,
+                    pa.field(col_name_int64, pa.int64()),
+                    pc.cast(table[column], pa.int64()),
+                )
+                for filter_value in partition_ranges:
+                    partitioned_tables[partition_ranges.index(filter_value)].append(
+                        table_new.filter(
+                            (pc.field(col_name_int64) <= pc.scalar(filter_value))
+                        )
+                    )
+                    if filter_value == partition_ranges[-1]:
+                        partitioned_tables[partition_ranges.index(filter_value)].append(
+                            table_new.filter(
+                                (pc.field(col_name_int64) > pc.scalar(filter_value))
+                            )
+                        )
+            # TODO(rootliu) set optimal or max number of records per file to defer the performance degradation due to too many small files
+            range_table_length = 0
+            range_deltas: List[Delta] = []
+            for _, value in partitioned_tables.items():
+                if len(value) > 0:
+                    range_table = pa.concat_tables(value)
+                    if len(range_table) > 0:
+                        range_table_length += len(range_table)
+                        range_delta = deltacat_storage.stage_delta(
+                            range_table,
+                            destination_partition,
+                            content_type=ContentType.PARQUET,
+                        )
+                        range_deltas.append(range_delta)
+
+            assert (
+                range_table_length == total_record_count
+            ), f"Repartitioned table should have the same number of records {range_table_length} as the original table {total_record_count}"
+            return RePartitionResult(
+                range_deltas=range_deltas,
+            )
+    else:
+        # Other repartition types, e.g., hash, key, list, column, etc are not supported yet
+        raise NotImplementedError(
+            f"Repartition type {repartition_type} is not supported."
+        )
```

Comment:
We definitely need a doc on what each repartition means and why they are required.